Add Libre Hardware Monitor integration

Sab44 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The current Open Hardware Monitor legacy integration is for a software that has not been updated in more than 4 years. It's successor is [Libre Hardware Monitor](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor).

While the legacy integration technically also works for Libre Hardware Monitor, it has many issues and is not maintained.

This integration has been rewritten from scratch following the best practices from the current developer docs, providing:

- config flow
- ~configurable scan interval~
- ~add only user selected hardware devices~
- entities grouped by hardware device
- ~reconfigure anytime via UI reconfigure flow~
- robust against hardware changes on the host system
- much improved code complexity and readability
- extensive test coverage
- quality scale: silver

In my opinion, we should deprecate the legacy integration and redirect users to this one.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: n/a
- This PR is related to issue: n/a
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/37941
- Link to developer documentation pull request: n/a
- Link to frontend pull request: n/a

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Sab44 committed:
add new librehardwaremonitor integration as successor to openhardwaremonitor

tr4nt0r reviewed (changes requested):
There are a few things you have to consider for a new integration. First of all, the PR scope has to be kept small. Only implement the user config flow, the reconfigure flow can be added in a follow-up PR.
Also variable/configurable scan interval is not allowed in core integrations.

Sab44 commented:
Hi @tr4nt0r thanks for the feedback.

Regarding the reconfigure flow: sure I can take this out and submit it in a follow-up PR.

Regarding the configurable scan interval: could you point me to the developer docs where it is outlined that it cannot be configurable? I found the [custom polling interval](https://www.home-assistant.io/common-tasks/general/#defining-a-custom-polling-interval) docs before, but it does not really say that a configurable scan interval is disallowed. 
I just want to note that I did set limitations to adhere to HA's minimum polling interval of 5 seconds. I feel like for a System Monitor it is vital to allow the user to specify their desired polling interval - ideally in a intuitive way during integration setup. But of course, if this is not desired then I can set a sensible default value and redirect to the custom polling interval docs on the integration docs.

joostlek commented:
Please note that someone is taking on the work on open hardware monitor and there's a PR for that at the moment.

Is the API of both applications the same?

Sab44 commented:
Hi @joostlek 
wow, what are the odds after years of legacy status... Not sure what to make of this.

Here's my thought process when I started working on implementing this integration from scratch:
- Open Hardware Monitor is legacy software, nobody should use it anymore. Instead they should use Libre Hardware Monitor. The integration name should reflect this, guiding new users intuitively to the right integration.
- The new integration should reflect the current state of development i.e. fulfill at least all the bronze quality requirements.

After a quick look at the other PR for `openhardwaremonitor`, in comparison this new integration provides:
- re-implementation instead of re-factor, avoiding pitfalls that are present in the old integration (and still are there after the re-factor, e.g. if hardware exposed by LHM changes, it will break the integration as sensors are based on indices that will no longer be present/valid)
- allow users to select which devices they want to add to Home Assistant
- high unit test coverage
- updated documentation & higher quality icon

Regarding the API, Libre Hardware Monitor seems to be backwards compatible, which is why the current `openhardwaremonitor` integration does technically still work. However, I have not tested whether my implementation will work with Open Hardware Monitor installations, as I don't think this is a valid use case anymore. Presumably it should work though.

joostlek commented:
I am also not saying which one is better, PRs should be well scoped, so I don't think comparing PRs is fair here, as new integrations are always bigger and contain more feature than an integration that is trying to modernize.

joostlek commented:
cc @lazytarget just as an FYI.

We just discussed this in the architecture meeting, and we think its fine to have them co-exist.

Sab44 commented:
Glad to hear that, thanks for clarifying and sorry for the hassle.

If someone could give me a quick hint why the tests on the CI are complaining about missing `data_description` translations when to my knowledge those are optional, I'd appreciate it. What am I missing there?

joostlek commented:
They are not optional for the quality_scale rule :)

LazyTarget commented:
> wow, what are the odds after years of legacy status... Not sure what to make of this.
@Sab44 
What ARE the odds?! 😄 

Cool to see that we both identify the need for an update, and interesting to see our different takes (even not all to different)
Really great job! I see you also put some hours into it.
I will see if can test yours out tomorrow

LazyTarget commented:
> I am also not saying which one is better, PRs should be well scoped, so I don't think comparing PRs is fair here, as new integrations are always bigger and contain more feature than an integration that is trying to modernize.
> 
> 
> We just discussed this in the architecture meeting, and we think its fine to have them co-exist.

@joostlek 
Sounds good. Should I revisit my PR a bit and make sure avoid breaking changes as much as possible?
Like continue supporting the YAML configuration?

joostlek commented:
With the migration to the UI we should import the YAML config and that is an expected breaking change

Sab44 commented:
> @Sab44
> What ARE the odds?! 😄

Yep, I guess it really was time to finally update this as LHM is such a nice tool and the legacy integration was increasingly becoming a pain point. 
Thank you too for your work and improvements! Let me know if you find any problems during your tests.

Sab44 commented:
There were a couple of code coverage warnings which are now resolved, so this is now ready for review.
If there is anything left to do from my side, please let me know.

lazytarget reviewed (changes requested):
Like mentioned in the other comment, In my case Libre Open Hardware Monitor is not able to get any sensors from my motherboard, which failed the startup.

Everything else looks and works great, truly good job!

joostlek commented on homeassistant/components/librehardwaremonitor/quality_scale.yaml:
```diff
@@ -0,0 +1,32 @@
+rules:
+  # Bronze
+  action-setup:
+    status: exempt
+    comment: |
+      No custom actions are defined.
+  appropriate-polling: done
+  brands: done
+  common-modules: done
+  config-flow-test-coverage: done
+  config-flow: done
+  dependency-transparency: done
+  docs-actions:
+    status: exempt
+    comment: |
+      No custom actions are defined.
+  docs-high-level-description: done
+  docs-installation-instructions: done
+  docs-removal-instructions: done
+  entity-event-setup:
+    status: exempt
+    comment: |
+      No explicit event subscriptions.
+  entity-unique-id: done
+  has-entity-name: done
+  runtime-data: done
+  test-before-configure: done
+  test-before-setup:
+    status: exempt
+    comment: |
+      Device is expected to be offline most of the time, but needs to connect quickly once available.
+  unique-config-entry: done
```
Let's add all the rules and evaluate them

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,98 @@
+    async_add_entities(sensor_entities)
+
+
+class LibreHardwareMonitorSensor(CoordinatorEntity, SensorEntity):
```
```suggestion
class LibreHardwareMonitorSensor(CoordinatorEntity[LibreHardwareMonitorCoordinator], SensorEntity):
```

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,98 @@
+        """Initialize an LibreHardwareMonitor sensor."""
+        super().__init__(coordinator)
+
+        self._coordinator = coordinator
```
already stored in `self.coordinator`

Sab44 replied:
Removed.

Sab44 committed:
move sensor parsing logic to librehardwaremonitor api package; add more types

Sab44 committed:
remove configurable scan interval

Sab44 committed:
update quality scale with complete checklist

Sab44 committed:
remove custom backoff time when host unreachable

Sab44 committed:
remove manual device selection and reconfiguration flow

Sab44 committed:
use device registry as source of truth for active devices

Sab44 committed:
fulfill strict typing quality scale requirement

Sab44 commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,145 @@
+
+            all_sensors_for_device = self._flatten_sensors(main_device)
+            for sensor in all_sensors_for_device:
+                sensor_id = "-".join(sensor[LHM_SENSOR_ID].split("/")[1:]).replace(
```
Thanks for checking, good catch. I did not expect there to be devices without both `SensorId` and `Children`. For me, the motherboard does provide quite some sensors, so your motherboard is probably a device that is not (yet) supported by LHM.
Anyways, it is fixed now, but the relevant code has moved to the api package. Thanks again!

Sab44 commented on homeassistant/components/librehardwaremonitor/__init__.py:
```diff
@@ -0,0 +1,28 @@
+PLATFORMS = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
```
Done.

Sab44 commented on homeassistant/components/librehardwaremonitor/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+                vol.Required(CONF_PORT, default=defaults[CONF_PORT]): int,
+                vol.Optional(
+                    CONF_SCAN_INTERVAL, default=defaults[CONF_SCAN_INTERVAL]
+                ): vol.All(vol.Coerce(int), vol.Range(min=5, max=30)),
```
Done.

Sab44 commented on homeassistant/components/librehardwaremonitor/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            step_id="user", data_schema=config_data_schema, errors=errors
+        )
+
+    async def async_step_select_devices(
```
Fair question. As Libre Hardware Monitor allows selection in itself of which Hardware devices to monitor, I removed the additional selection from the integration here.

Sab44 commented on homeassistant/components/librehardwaremonitor/config_flow.py:
```diff
@@ -0,0 +1,170 @@
+            step_id="select_devices", data_schema=devices_data_schema
+        )
+
+    async def async_step_reconfigure(
```
Removed the reconfigure flow.

Sab44 commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,145 @@
+LHM_VALUE = "Value"
+
+
+class LibreHardwareMonitorCoordinator(DataUpdateCoordinator):
```
Done.

Sab44 commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,145 @@
+        except LibreHardwareMonitorConnectionError as err:
+            self.update_interval = timedelta(
+                seconds=RETRY_INTERVAL_WHEN_HOST_UNREACHABLE
+            )
```
As the scan interval is no longer configurable I removed this logic. Presumably if someone sets a high polling frequency via automations, they can check themselves whether the device is online in the first place.

The idea behind this was, that the monitored computer is expected to be offline, potentially even most of the time. If the user had configured a scan interval of e.g. 5 seconds but the computer goes down while HA is running, then the integration would keep polling unsuccessfully every 5 seconds which I found to be unnecessary. So the idea was, poll every 5 seconds if device is online, poll every 30 seconds while device is offline and revert to 5 seconds once it becomes online again.
(You're right of course that if HA starts up and the computer is offline, HA uses its in-built retry strategy with increasing back-off duration. But as it does not handle the case where HA is running already, it did not seem sufficient.)

Sab44 commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,145 @@
+                    device_type=device_type,
+                    sensor_id=sensor_id,
+                )
+                sensors_data[sensor_id] = sensor_data
```
I agree. This is implemented now, the library will return the data parsed to a dataclass.

Sab44 commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,98 @@
+
+        self._attr_name = sensor_data.name
+        self.value = sensor_data.value
+        self.attributes = {
+            STATE_MIN_VALUE: self._format_number_value(sensor_data.min),
+            STATE_MAX_VALUE: self._format_number_value(sensor_data.max),
+        }
```
This is a feature of Libre Hardware Monitor, it keeps tracking those values so we just add them here since they are provided by the API. Users might want to keep track of those values or use them for automations etc.

![Screenshot 2025-04-17 at 08 46 08](https://github.com/user-attachments/assets/e0b57329-7f72-4f25-bcbf-7dcd6802a71e)

Sab44 commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,98 @@
+    _attr_state_class = SensorStateClass.MEASUREMENT
+    _attr_has_entity_name = True
+
+    def __init__(self, coordinator, sensor_data):
```
Done.

Sab44 commented:
@joostlek I have addressed all comments, did not resolve them yet so you can check them again. I structured all changes into hopefully easily reviewable commits, but I can of course squash them if desired. The docs PR is also updated to reflect changes made here.

Sab44 committed:
bump librehardwaremonitor-api dependency

joostlek commented on homeassistant/components/librehardwaremonitor/config_flow.py:
```diff
@@ -0,0 +1,70 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        config_data_schema = vol.Schema(
+            {
+                vol.Required(CONF_HOST, default=DEFAULT_HOST): str,
+                vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+            }
+        )
```
this can become a constant

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/config_flow.py:
```diff
@@ -0,0 +1,70 @@
+        errors = {}
+
+        if user_input is not None:
+            await self.async_set_unique_id(
+                f"{user_input[CONF_HOST]}:{user_input[CONF_PORT]}"
+            )
```
host and port is not a valid unique id, we should use self._async_abort_entries_match here

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/config_flow.py:
```diff
@@ -0,0 +1,70 @@
+            try:
+                _ = (await api.get_data()).main_device_names
+            except LibreHardwareMonitorConnectionError:
+                _LOGGER.debug(
+                    "ConnectionError: Is LibreHardwareMonitor running and the web server option enabled?"
+                )
+                errors["base"] = "cannot_connect"
+            except LibreHardwareMonitorNoDevicesError:
+                _LOGGER.debug(
+                    "NoDevicesError: LibreHardwareMonitor did not return any devices"
+                )
```
Since we only have these 2 cases, the debug lines don't add much

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+        host = config_entry.data[CONF_HOST]
+        port = config_entry.data[CONF_PORT]
+        self._api = LibreHardwareMonitorClient(host, port)
+        device_entries = dr.async_entries_for_config_entry(
+            registry=dr.async_get(self.hass), config_entry_id=config_entry.entry_id
+        )
+        self._previous_devices = {
+            device.name for device in device_entries if device.name is not None
+        }
```
I would recommend to split off the stale and dynamic devices for now

Sab44 replied:
I understand the scope of the PR has to be reasonably small, that why I removed the reconfigure flow, the device selection and more.
In this case, I'm hesitant to remove it. Having the devices in HA reflect the actual devices exposed by LHM is a very nice feature to have and a real user-facing benefit. This functionality is of course covered by unit tests.

joostlek replied:
Oh I agree that its a nice feature, but I think we can be a bit quicker with reviewing without this, and then we can add this in a follow-up

Sab44 replied:
I'd really like to keep it, hope it doesn't add too much review effort. As there is no urgency for this PR I think we can afford to keep it and avoid device inconsistencies in users' configs that we probably will have to clean up later.

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+
+        return lhm_data
+
+    async def _async_refresh(
+        self,
+        log_failures: bool = True,
+        raise_on_auth_failed: bool = False,
+        scheduled: bool = False,
+        raise_on_entry_error: bool = False,
+    ) -> None:
+        # we don't expect the computer to be online 24/7 so we don't want to log a connection loss as an error
+        await super()._async_refresh(
+            False, raise_on_auth_failed, scheduled, raise_on_entry_error
+        )
```
Would you be willing to move this to a followup PR?

Sab44 replied:
One of the reasons I developed this integration is that the old Open Hardware Monitor integration would spam my HA logs with warnings.
If your concern is with overwriting the `_async_refresh` method, I could also adapt the `_async_update_data` logic to not raise `UpdateFailed` exceptions which should therefore not cause any undesired logs.

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,108 @@
+    """Set up the LibreHardwareMonitor platform."""
+    lhm_coordinator: LibreHardwareMonitorCoordinator = config_entry.runtime_data
+
+    sensor_entities = [
+        LibreHardwareMonitorSensor(lhm_coordinator, sensor_data)
+        for sensor_data in lhm_coordinator.data.sensor_data.values()
+    ]
+
+    async_add_entities(sensor_entities)
```
can be merged

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,108 @@
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up the LibreHardwareMonitor platform."""
+    lhm_coordinator: LibreHardwareMonitorCoordinator = config_entry.runtime_data
```
```suggestion
    lhm_coordinator = config_entry.runtime_data
```

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,108 @@
+            STATE_MIN_VALUE: self._format_number_value(sensor_data.min),
+            STATE_MAX_VALUE: self._format_number_value(sensor_data.max),
+        }
+        self._unit_of_measurement: str | None = sensor_data.unit
```
```suggestion
        self._attr_native_unit_of_measurement = sensor_data.unit
```

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,108 @@
+
+        self.async_write_ha_state()
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement."""
+        return self._unit_of_measurement
```
```suggestion
```

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,108 @@
+
+        self._attr_name: str = sensor_data.name
+        self.value: str | None = sensor_data.value
+        self.attributes: dict[str, str] = {
```
```suggestion
        self._attr_extra_state_attributes = {
```

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,108 @@
+
+        # Hardware device
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, sensor_data.device_name)},
```
What is the device name in this case? Is it truly unique?

Sab44 replied:
The device name is the actual hardware device the sensor belongs to. So for example:
- sensor 1: "temperature-gpu-core"
- sensor 2: "temperature-gpu-hotspot"
- sensor 3: "load-gpu-core"
will all share the same device name, e.g. something like "NVIDIA RTX 4080".
They will be grouped accordingly in the UI.

joostlek replied:
Do we have anything that looks like a unique identifier?

Sab44 replied:
Actually there is a unique identifier. I have adapted the library to expose it and made the necessary changes to the integration's code. It should be fine now, even if someone has the same SSD twice in their PC.

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,108 @@
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, sensor_data.device_name)},
+            name=sensor_data.device_name,
+            manufacturer="Hardware",
```
who is the manufactuerer?

Sab44 replied:
It's not possible to get the manufacterer from LHM, I removed the "Hardware" placeholder.

Sab44 committed:
remove debug lines from config flow

Sab44 committed:
small refactor; resolve minor comments

Sab44 committed:
log connection error

Sab44 committed:
extract config schema into constant

Sab44 committed:
remove dummy manufacterer name

Sab44 committed:
replace unique id with async_abort_entries_match

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,102 @@
+    """Set up the LibreHardwareMonitor platform."""
+    lhm_coordinator = config_entry.runtime_data
+
+    async_add_entities(
+        [
+            LibreHardwareMonitorSensor(lhm_coordinator, sensor_data)
+            for sensor_data in lhm_coordinator.data.sensor_data.values()
+        ]
+    )
```
```suggestion
    async_add_entities(
        LibreHardwareMonitorSensor(lhm_coordinator, sensor_data)
        for sensor_data in lhm_coordinator.data.sensor_data.values()
    )
```

joostlek commented on tests/components/librehardwaremonitor/__init__.py:
```diff
@@ -0,0 +1,25 @@
+
+async def init_integration(hass: HomeAssistant) -> MockConfigEntry:
+    """Set up the LibreHardwareMonitor integration in Home Assistant."""
+    mock_entry = MockConfigEntry(
+        domain=DOMAIN,
+        title="192.168.0.20:8085",
+        unique_id="192.168.0.20:8085",
+        data=VALID_CONFIG,
+    )
```
I would recommend to make the mock config entry a test fixture as you will use this in more places

Sab44 replied:
Done.

joostlek commented on tests/components/librehardwaremonitor/conftest.py:
```diff
@@ -0,0 +1,290 @@
+        yield client
+
+
+LHM_SAMPLE_DATA = LibreHardwareMonitorData(
```
How do generally map this from JSON? If you can do the same here you could use a json fixture file

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,136 @@
+    mock_lhm_client.reset_mock()
+
+
+async def test_sensor_state_is_unknown_when_lhm_indicates_missing_value(
+    hass: HomeAssistant, mock_lhm_client: AsyncMock
+) -> None:
+    """Test sensor state is unknown when no value is present."""
+    entity_id = "sensor.msi_mag_b650m_mortar_wifi_ms_7d76_system_fan_1_fan"
+    await init_integration(hass)
+
+    state = hass.states.get(entity_id)
+
+    assert state
+    assert state.state == STATE_UNKNOWN
+    assert state.name == "MSI MAG B650M MORTAR WIFI (MS-7D76) System Fan #1 Fan"
```
Then you can also get rid of all the `state.name == ....` in the other tests

Sab44 replied:
Good point! Done.

Sab44 committed:
remove redundant list designation for sensor entities

Sab44 committed:
rework unit tests
* add mock_config_entry fixture
* combine error config flow tests
* remove unnecessary config flow test

Sab44 committed:
patch async_setup_entry to avoid leaking setups

Sab44 committed:
add snapshot test to assert created entities state

Sab44 commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+
+    assert result
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "user"
```
Removed.

Sab44 commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+    )
+
+    assert result
+    assert result["errors"] == {"base": "no_devices"}
```
Done.

Sab44 commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+
+async def test_lhm_server_already_exists(hass: HomeAssistant) -> None:
+    """Test we only allow a single entry per server."""
+    await init_integration(hass)
```
Done.

Sab44 commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,85 @@
+    assert config_entry.data == {
+        CONF_HOST: VALID_CONFIG[CONF_HOST],
+        CONF_PORT: VALID_CONFIG[CONF_PORT],
+    }
```
Done.

Sab44 commented on tests/components/librehardwaremonitor/test_config_flow.py:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+class LibreHardwareMonitorCoordinator(DataUpdateCoordinator[LibreHardwareMonitorData]):
+    """Class to manage fetching LibreHardwareMonitor data."""
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
```
```suggestion
    def __init__(self, hass: HomeAssistant, config_entry: LibreHardwareMonitorConfigEntry) -> None:
```

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+
+        return lhm_data
+
+    async def _async_refresh(
+        self,
+        log_failures: bool = True,
+        raise_on_auth_failed: bool = False,
+        scheduled: bool = False,
+        raise_on_entry_error: bool = False,
+    ) -> None:
+        # we don't expect the computer to be online 24/7 so we don't want to log a connection loss as an error
+        await super()._async_refresh(
+            False, raise_on_auth_failed, scheduled, raise_on_entry_error
+        )
```
Okay so the issue with this is, even if we can't connect, we should render the entities unavailable. And I kinda think that having an error in the logs is part of that experience (as in, you don't know exactly _why_ it got disconnected)

Sab44 replied:
I'm sure there are use-cases where a continuous monitoring is desired and any connection loss should be considered an error. As this tool is a monitoring tool for Windows PCs, I'd argue it is more used in a non-24/7 scenario. Here's my use-case, which I believe is rather common:

- Windows Gaming PC that is shut down / in sleep most of the time
- When turned on / actually used for gaming I want to monitor my hardware
- After being done I turn the PC off
- This can happen multiple times a day
- An error log every time this happens can be annoying

My suggestion would be to extend the integration docs with information about this behavior and provide an example automation that checks for a sensor to become unavailable and triggers a notification if the user requires this functionality. What do you think?

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+        self, detected_devices: list[str]
+    ) -> None:
+        """Handle device changes by deleting devices from / adding devices to Home Assistant."""
+        if self._previous_devices == set(detected_devices) or self.config_entry is None:
```
why would the config entry by None?

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+        if self._previous_devices == set(detected_devices) or self.config_entry is None:
+            return
+
+        if self.data is None:
+            self._previous_devices = set(detected_devices)  # type: ignore[unreachable]
+            return
```
Why do we recalculate the previous devices?

Sab44 replied:
During the very first coordinator update, we expect the devices to change (from <no devices> to whatever x devices we find). So we don't consider this a change in the existing devices. There is also a unit test for this.

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+                        remove_config_entry_id=self.config_entry.entry_id,
+                    )
+
+        if new_devices := list(set(detected_devices) - self._previous_devices):
+            _LOGGER.info("New Device(s) added, reloading integration: %s", new_devices)
+            self.hass.config_entries.async_schedule_reload(self.config_entry.entry_id)
```
Let's keep this particular part out for now, I think we can improve this better

Sab44 replied:
I removed the reloading of the integration, but kept the log statement so it's at least communicated that there's a change in devices detected.

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+            return
+
+        if orphaned_devices := list(self._previous_devices - set(detected_devices)):
+            _LOGGER.info(
+                "Devices no longer available, will be removed: %s", orphaned_devices
+            )
```
Let's not log this on info

joostlek commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,108 @@
+            self._previous_devices = set(detected_devices)  # type: ignore[unreachable]
+            return
+
+        if orphaned_devices := list(self._previous_devices - set(detected_devices)):
```
why do we `list` this?

Sab44 replied:
Reworked this part to be based off the just introduced device ids instead of names and changed the implementation to consistently work with sets and remove all `list` occurrences from here.

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,100 @@
+        else:
+            self.value = None
+
+        self.async_write_ha_state()
```
```suggestion
        super._handle_coordinator_update()
```

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,100 @@
+            return self._format_number_value(self.value)
+        return None
+
+    @property
+    def extra_state_attributes(self) -> dict[str, str]:
+        """Return the state attributes of the entity."""
+        return self._attr_extra_state_attributes
```
```suggestion
```

Sab44 replied:
Done.

joostlek commented on homeassistant/components/librehardwaremonitor/sensor.py:
```diff
@@ -0,0 +1,100 @@
+            model=sensor_data.device_type,
+        )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        if sensor_data := self.coordinator.data.sensor_data.get(self._sensor_id):
+            self.value = sensor_data.value
+            self._attr_extra_state_attributes = {
+                STATE_MIN_VALUE: self._format_number_value(sensor_data.min),
+                STATE_MAX_VALUE: self._format_number_value(sensor_data.max),
+            }
+        else:
+            self.value = None
+
+        self.async_write_ha_state()
+
+    @property
+    def native_value(self) -> str | None:
+        """Return the state of the device."""
+        if self.value is not None and self.value != "-":
+            return self._format_number_value(self.value)
+        return None
+
+    @property
+    def extra_state_attributes(self) -> dict[str, str]:
+        """Return the state attributes of the entity."""
+        return self._attr_extra_state_attributes
```
I think I would actually recommend to either use `_handle_coordinator_update` or just use properties and extract the data on the run

Sab44 committed:
provide unique device id for hardware device

Sab44 committed:
various smaller PR review improvements

Sab44 committed:
check changes in devices based on device ids

Sab44 committed:
do not automatically add new devices

Sab44 committed:
remove config entry None check

joostlek commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+        DOMAIN, context={"source": SOURCE_USER}, data=VALID_CONFIG
+    )
+
+    assert result
```
No need to assert result everytime, it will be there

joostlek commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+    )
+
+    assert result
+    assert result["errors"] == {"base": "cannot_connect"}
+    assert result["step_id"] == "user"
```
Let's also assert the type of the result is form

joostlek commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+
+    mock_lhm_client.get_data.side_effect = LibreHardwareMonitorNoDevicesError()
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}, data=VALID_CONFIG
+    )
```
This is creating a new flow instead of re-using the old one

Sab44 replied:
Fixed it.

joostlek commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+
+    mock_lhm_client.get_data.side_effect = None
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}, data=VALID_CONFIG
+    )
```
idem

joostlek commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+    """Test we only allow a single entry per server."""
+    mock_config_entry.add_to_hass(hass)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}, data=VALID_CONFIG
+    )
```
split this one as well

joostlek commented on tests/components/librehardwaremonitor/test_coordinator.py:
So I personally strongly dislike coordinator tests. The coordinator is a tool, rather than the endgoal. As in, the coordinator results in behaviour in the entities states, or in this case, the available entities or devices. So instead this should result in more sensor tests, instead of this file.

Also, we should not create coordinators ourself, we want HA to create it

Sab44 replied:
I understand what you're saying.

But since I don't really have a stake in the architectural design that Home Assistant enforces for its integrations I'm not sure what you're aiming for with this info.

The [developer docs](https://developers.home-assistant.io/docs/integration_fetching_data/) clearly describe the use of a coordinator so that's what I followed under the impression of using a best-practice approach to integration development.
So is this just a philosophical debate or are you suggesting getting rid of the coordinator? If so, then please name some integrations that use the current coordinator-less best-practice and I will try to adapt the code.

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,132 @@
+from tests.common import MockConfigEntry, async_fire_time_changed
+
+
+async def test_sensors_are_created(
+    hass: HomeAssistant,
+    mock_lhm_client: AsyncMock,
+    mock_config_entry: MockConfigEntry,
+    freezer: FrozenDateTimeFactory,
+    snapshot: SnapshotAssertion,
+) -> None:
+    """Test sensors are created."""
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    initial_states = hass.states.async_all()
+    assert initial_states == snapshot(name="valid_sensor_data")
```
Please use the `snapshot_platform` helper for this.

Sab44 replied:
Done.

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,132 @@
+    snapshot: SnapshotAssertion,
+) -> None:
+    """Test sensors are created."""
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
```
you can put this in a function

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,132 @@
+    initial_states = hass.states.async_all()
+    assert initial_states == snapshot(name="valid_sensor_data")
+
+    mock_lhm_client.get_data.side_effect = LibreHardwareMonitorConnectionError()
+
+    freezer.tick(timedelta(DEFAULT_SCAN_INTERVAL))
+    async_fire_time_changed(hass)
+    await hass.async_block_till_done()
+
+    unavailable_states = hass.states.async_all()
+    assert all(state.state == STATE_UNAVAILABLE for state in unavailable_states)
+
+    mock_lhm_client.get_data.side_effect = None
+
+    freezer.tick(timedelta(DEFAULT_SCAN_INTERVAL))
+    async_fire_time_changed(hass)
+    await hass.async_block_till_done()
```
I would argue that this is a different test and that we should split it

Sab44 replied:
Extracted it into its own test function.

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,132 @@
+    await hass.async_block_till_done()
+
+    recovered_states = hass.states.async_all()
+    assert recovered_states == snapshot(name="valid_sensor_data")
```
Snapshots are cool. They help with fixating data and tracking change. What they are bad at, is tracking change in the same test. I would rather see a harder assertion here, so we don't hide these details in the ambr file

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,132 @@
+    assert state.state != STATE_UNAVAILABLE
+    assert state.state == "39.4"
+
+    updated_data = copy.deepcopy(mock_lhm_client.get_data.return_value)
+    updated_data.sensor_data["amdcpu-0-temperature-3"].value = "42,1"
+    mock_lhm_client.get_data.return_value = updated_data
```
I don't thin you have to deep copy

Sab44 replied:
Actually it was necessary, but I rewrote this nonetheless as the underlying issue was with mutability of the state so I introduced immutable types in the library.

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,132 @@
+    assert state.state != STATE_UNAVAILABLE
+    assert state.state == "42.1"
+
+    mock_lhm_client.reset_mock()
```
we dont have to reset the mock right

joostlek commented on tests/components/librehardwaremonitor/test_sensor.py:
```diff
@@ -0,0 +1,132 @@
+    mock_lhm_client.reset_mock()
+
+
+async def test_sensor_state_is_unknown_when_lhm_indicates_missing_value(
+    hass: HomeAssistant, mock_lhm_client: AsyncMock, mock_config_entry: MockConfigEntry
+) -> None:
+    """Test sensor state is unknown when no value is present."""
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    entity_id = "sensor.msi_mag_b650m_mortar_wifi_ms_7d76_system_fan_1_fan"
+
+    state = hass.states.get(entity_id)
+
+    assert state
+    assert state.state == STATE_UNKNOWN
```
isnt this already tested by the snapshot test?

Sab44 replied:
You're right, I removed it.

Sab44 committed:
add underscores to component name

Sab44 committed:
implement suggested changes to coordinator

Sab44 committed:
remove unique id from conftest setup

Sab44 committed:
improve config flow tests

Sab44 committed:
update librehardwaremonitor-api dependency

Sab44 committed:
implement test review suggestions

Sab44 committed:
add json fixture and parse for test data

Sab44 commented on homeassistant/components/librehardwaremonitor/__init__.py:
Refactored the code now to add the underscores.

Sab44 committed:
change dict to immutable type in coordinator

Sab44 commented on homeassistant/components/librehardwaremonitor/coordinator.py:
```diff
@@ -0,0 +1,123 @@
+
+        if self.data is None:
+            # initial update during integration startup
+            self._previous_devices = detected_devices  # type: ignore[unreachable]
```
mypy complains about the statement being unreachable here. Not sure why tbh. The unit tests certainly prove it's reachable.

Sab44 commented on tests/components/librehardwaremonitor/test_config_flow.py:
```diff
@@ -0,0 +1,99 @@
+    assert mock_config_entry.data == {
+        CONF_HOST: VALID_CONFIG[CONF_HOST],
+        CONF_PORT: VALID_CONFIG[CONF_PORT],
+    }
```
Done.

Sab44 committed:
re-add missing value into fixture and update snapshot

Sab44 commented on homeassistant/components/libre_hardware_monitor/coordinator.py:
```diff
@@ -0,0 +1,130 @@
+        if previous_device_ids == detected_device_ids:
+            return
+
+        if self.data is None:
+            # initial update during integration startup
+            self._previous_devices = detected_devices  # type: ignore[unreachable]
+            return
```
See https://github.com/home-assistant/core/pull/140449#discussion_r2219474053

Sab44 commented:
> I think the test_sensor should have tests that tests that the logic works as expected, as we can test the state machine that the logic in the coordinator works

Ok so just to summarize, if I understand correctly this is how I'll proceed:
- we keep the coordinator itself
- we try to keep the coordinator tests as lean as possible
- we move as many tests as possible from the coordinator tests to the sensor tests

Correct?

Sab44 committed:
transform coordinator tests into sensor tests

Sab44 committed:
remove coordinator tests; extend sensor test for errors

Sab44 commented on tests/components/libre_hardware_monitor/test_coordinator.py:
```diff
@@ -0,0 +1,43 @@
+    coordinator = LibreHardwareMonitorCoordinator(hass, mock_config_entry)
+
+    with pytest.raises(UpdateFailed):
+        await coordinator._async_update_data()
```
I have removed the coordinator tests altogether now. Extended the sensor tests to cover the same error scenarios.
